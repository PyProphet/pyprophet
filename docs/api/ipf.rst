Inference of Peptidoforms Documentation
==========================

.. automodule:: pyprophet.ipf
   :no-members:
   :no-inherited-members:

.. currentmodule:: pyprophet.ipf

.. autosummary::
   :nosignatures:
   :toctree: generated/
   :template: class.rst

   infer_peptidoforms
   peptidoform_inference
   precursor_inference
   apply_bm
   prepare_transition_bm
   transfer_confident_evidence_across_runs
   prepare_precursor_bm
   compute_model_fdr