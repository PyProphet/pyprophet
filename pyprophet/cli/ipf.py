import click
from loguru import logger

from .util import write_logfile, measure_memory_usage_and_time
from .._config import IPFIOConfig
from ..ipf import infer_peptidoforms
from ..glyco.glycoform import infer_glycoforms


# IPF
@click.command()
# File handling
@click.option(
    "--in",
    "infile",
    required=True,
    type=click.Path(exists=True),
    help="PyProphet input file. Valid formats are .osw, .parquet (produced by export parquet)",
)
@click.option(
    "--out",
    "outfile",
    type=click.Path(exists=False),
    help="PyProphet output file. Valid formats are .osw, .parquet. Must be the same format as input file.",
)
# IPF parameters
@click.option(
    "--ipf_ms1_scoring/--no-ipf_ms1_scoring",
    default=True,
    show_default=True,
    help="Use MS1 precursor data for IPF.",
)
@click.option(
    "--ipf_ms2_scoring/--no-ipf_ms2_scoring",
    default=True,
    show_default=True,
    help="Use MS2 precursor data for IPF.",
)
@click.option(
    "--ipf_h0/--no-ipf_h0",
    default=True,
    show_default=True,
    help="Include possibility that peak groups are not covered by peptidoform space.",
)
@click.option(
    "--ipf_grouped_fdr/--no-ipf_grouped_fdr",
    default=False,
    show_default=True,
    help="[Experimental] Compute grouped FDR instead of pooled FDR to better support data where peak groups are evaluated to originate from very heterogeneous numbers of peptidoforms.",
)
@click.option(
    "--ipf_max_precursor_pep",
    default=0.7,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored precursors in IPF.",
)
@click.option(
    "--ipf_max_peakgroup_pep",
    default=0.7,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored peak groups in IPF.",
)
@click.option(
    "--ipf_max_precursor_peakgroup_pep",
    default=0.4,
    show_default=True,
    type=float,
    help="Maximum BHM layer 1 integrated precursor peakgroup PEP to consider in IPF.",
)
@click.option(
    "--ipf_max_transition_pep",
    default=0.6,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored transitions in IPF.",
)
@click.option(
    "--propagate_signal_across_runs/--no-propagate_signal_across_runs",
    default=False,
    show_default=True,
    help="Propagate signal across runs (requires running alignment).",
)
@click.option(
    "--ipf_max_alignment_pep",
    default=1.0,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for good alignments.",
)
@click.option(
    "--across_run_confidence_threshold",
    default=0.5,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for propagating signal across runs for aligned features.",
)
@click.pass_context
@measure_memory_usage_and_time
@logger.catch(reraise=True)
def ipf(
    ctx,
    infile,
    outfile,
    ipf_ms1_scoring,
    ipf_ms2_scoring,
    ipf_h0,
    ipf_grouped_fdr,
    ipf_max_precursor_pep,
    ipf_max_peakgroup_pep,
    ipf_max_precursor_peakgroup_pep,
    ipf_max_transition_pep,
    propagate_signal_across_runs,
    ipf_max_alignment_pep,
    across_run_confidence_threshold,
):
    """
    Infer peptidoforms after scoring of MS1, MS2 and transition-level data.
    """

    # Check to see if we're calling from the previous deprecated command call
    if ctx.command.deprecated:
        logger.warning(
            "The 'pyprophet ipf' command is deprecated and will be removed in future versions. Use 'pyprophet infer peptidoform' instead."
        )

    if outfile is None:
        outfile = infile
    else:
        outfile = outfile

    config = IPFIOConfig.from_cli_args(
        infile,
        outfile,
        1,  # Subsample ratio is not applicable for IPF
        "ipf",  # Level is not applicable for IPF
        "ipf",
        ipf_ms1_scoring,
        ipf_ms2_scoring,
        ipf_h0,
        ipf_grouped_fdr,
        ipf_max_precursor_pep,
        ipf_max_peakgroup_pep,
        ipf_max_precursor_peakgroup_pep,
        ipf_max_transition_pep,
        propagate_signal_across_runs,
        ipf_max_alignment_pep,
        across_run_confidence_threshold,
    )
    write_logfile(
        ctx.obj["LOG_LEVEL"], f"{config.prefix}_pyp_ipf.log", ctx.obj["LOG_HEADER"]
    )
    infer_peptidoforms(config)


# Infer glycoforms
@click.command()
@click.option(
    "--in", "infile", required=True, type=click.Path(exists=True), help="Input file."
)
@click.option("--out", "outfile", type=click.Path(exists=False), help="Output file.")
@click.option(
    "--ms1_precursor_scoring/--no-ms1_precursor_scoring",
    default=True,
    show_default=True,
    help="Use MS1 precursor data for glycoform inference.",
)
@click.option(
    "--ms2_precursor_scoring/--no-ms2_precursor_scoring",
    default=True,
    show_default=True,
    help="Use MS2 precursor data for glycoform inference.",
)
@click.option(
    "--grouped_fdr/--no-grouped_fdr",
    default=False,
    show_default=True,
    help="[Experimental] Compute grouped FDR instead of pooled FDR to better support data where peak groups are evaluated to originate from very heterogeneous numbers of glycoforms.",
)
@click.option(
    "--max_precursor_pep",
    default=1,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored precursors.",
)
@click.option(
    "--max_peakgroup_pep",
    default=0.7,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored peak groups.",
)
@click.option(
    "--max_precursor_peakgroup_pep",
    default=1,
    show_default=True,
    type=float,
    help="Maximum BHM layer 1 integrated precursor peakgroup PEP to consider.",
)
@click.option(
    "--max_transition_pep",
    default=0.6,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored transitions.",
)
@click.option(
    "--use_glycan_composition/--use_glycan_struct",
    "use_glycan_composition",
    default=True,
    show_default=True,
    help="Compute glycoform-level FDR based on glycan composition or struct.",
)
@click.option(
    "--ms1_mz_window",
    default=10,
    show_default=True,
    type=float,
    help="MS1 m/z window in Thomson or ppm.",
)
@click.option(
    "--ms1_mz_window_unit",
    default="ppm",
    show_default=True,
    type=click.Choice(["ppm", "Da", "Th"]),
    help="MS1 m/z window unit.",
)
@click.option(
    "--propagate_signal_across_runs/--no-propagate_signal_across_runs",
    default=False,
    show_default=True,
    help="Propagate signal across runs (requires running alignment).",
)
@click.option(
    "--max_alignment_pep",
    default=1.0,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for good alignments.",
)
@click.option(
    "--across_run_confidence_threshold",
    default=0.5,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for propagating signal across runs for aligned features.",
)
def glycoform(
    infile,
    outfile,
    ms1_precursor_scoring,
    ms2_precursor_scoring,
    grouped_fdr,
    max_precursor_pep,
    max_peakgroup_pep,
    max_precursor_peakgroup_pep,
    max_transition_pep,
    use_glycan_composition,
    ms1_mz_window,
    ms1_mz_window_unit,
    propagate_signal_across_runs,
    max_alignment_pep,
    across_run_confidence_threshold,
):
    """
    Infer glycoforms after scoring of MS1, MS2 and transition-level data.
    """

    if outfile is None:
        outfile = infile

    infer_glycoforms(
        infile=infile,
        outfile=outfile,
        ms1_precursor_scoring=ms1_precursor_scoring,
        ms2_precursor_scoring=ms2_precursor_scoring,
        grouped_fdr=grouped_fdr,
        max_precursor_pep=max_precursor_pep,
        max_peakgroup_pep=max_peakgroup_pep,
        max_precursor_peakgroup_pep=max_precursor_peakgroup_pep,
        max_transition_pep=max_transition_pep,
        use_glycan_composition=use_glycan_composition,
        ms1_mz_window=ms1_mz_window,
        ms1_mz_window_unit=ms1_mz_window_unit,
        propagate_signal_across_runs=propagate_signal_across_runs,
        max_alignment_pep=max_alignment_pep,
        across_run_confidence_threshold=across_run_confidence_threshold,
    )
