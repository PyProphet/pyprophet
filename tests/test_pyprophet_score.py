from __future__ import print_function

import os
import subprocess
import shutil
import sys

import pandas as pd
import sqlite3

import pytest

from pyprophet.ipf import read_pyp_peakgroup_precursor

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += str(subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT))
    except subprocess.CalledProcessError as error:
        print(error, end="", file=sys.stderr)
        raise
    return stdout

def _run_pyprophet_tsv_to_learn_model(regtest, temp_folder, dump_result_files=False, parametric=False, pfdr=False, pi0_lambda=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.txt")
    shutil.copy(data_path, temp_folder)
    cmdline = "pyprophet score --pi0_method=smoother --pi0_lambda 0.4 0 0 --in=test_data.txt --test --ss_iteration_fdr=0.02"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_summary_stat.csv", sep=",", nrows=100).sort_index(axis=1),file=regtest)
    print(pd.read_csv("test_data_full_stat.csv", sep=",", nrows=100).sort_index(axis=1),file=regtest)
    print(pd.read_csv("test_data_scored.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    print(pd.read_csv("test_data_weights.csv", sep=",", nrows=100).sort_index(axis=1),file=regtest)

def _run_pyprophet_osw_to_learn_model_metabo(regtest, temp_folder, dump_result_files=False, ms1ms2=False, score_filter=False):
    os.chdir(temp_folder)
    data_path_metabo = os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path_metabo, temp_folder)

    # MS1-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms1 --test --ss_iteration_fdr=0.02"
    if score_filter:
        cmdline += " --ss_score_filter var_isotope_overlap_score,var_massdev_score,var_xcorr_coelution,var_isotope_correlation_score"

    # MS2-level
    if ms1ms2:
            cmdline += " score --in=test_data_compound.osw --level=ms1ms2 --test --ss_iteration_fdr=0.02"
            if score_filter:
                cmdline += " --ss_score_filter metabolomics"
    else:
        cmdline += " score --in=test_data_compound.osw --level=ms2 --test --ss_iteration_fdr=0.02"
        if score_filter:
            cmdline += " --ss_score_filter var_isotope_overlap_score,var_isotope_correlation_score,var_intensity_score,var_massdev_score,var_library_corr,var_norm_rt_score"

    stdout = _run_cmdline(cmdline)

    scored = sqlite3.connect("test_data_compound.osw")
    table = pd.read_sql_query("SELECT * FROM PYPROPHET_WEIGHTS ORDER BY score", scored)
    scored.close()

    print(table.head(100).sort_index(axis=1),file=regtest)

def _run_pyprophet_osw_to_learn_model(regtest, temp_folder, dump_result_files=False, parametric=False, pfdr=False, pi0_lambda=False, ms1ms2=False, xgboost=False, xgboost_tune=False, score_filter=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)

    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms1 --test --ss_iteration_fdr=0.02"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    if xgboost:
        cmdline += " --classifier=XGBoost"
    if xgboost_tune:
        cmdline += " --xgb_autotune"
    if score_filter:
        cmdline += " --ss_score_filter var_isotope_overlap_score,var_massdev_score,var_xcorr_coelution,var_isotope_correlation_score"

    # MS2-level
    if ms1ms2:
        cmdline += " score --in=test_data.osw --level=ms1ms2 --test --ss_iteration_fdr=0.02"
        if score_filter:
            cmdline += " --ss_score_filter var_ms1_isotope_overlap_score,var_ms1_massdev_score,var_ms1_xcorr_coelution,var_ms1_isotope_correlation_score,var_isotope_overlap_score,var_isotope_correlation_score,var_intensity_score,var_massdev_score,var_library_corr,var_norm_rt_score"
    else:
        cmdline += " score --in=test_data.osw --level=ms2 --test --ss_iteration_fdr=0.02"
        if score_filter:
            cmdline += " --ss_score_filter var_isotope_overlap_score,var_isotope_correlation_score,var_intensity_score,var_massdev_score,var_library_corr,var_norm_rt_score"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    if xgboost:
        cmdline += " --classifier=XGBoost"
    if xgboost_tune:
        cmdline += " --xgb_autotune"

    # transition-level
    cmdline += " score --in=test_data.osw --level=transition --test --ss_iteration_fdr=0.02"

    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    if xgboost:
        cmdline += " --classifier=XGBoost"
    if xgboost_tune:
        cmdline += " --xgb_autotune"

    stdout = _run_cmdline(cmdline)

    table = read_pyp_peakgroup_precursor("test_data.osw", 1.0, True, True)

    print(table.head(100).sort_index(axis=1),file=regtest)

def test_tsv_0(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True)

def test_tsv_1(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True, True)

def test_tsv_2(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True, False, True)

def test_tsv_3(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True, False, False, "0.3 0.55 0.05")

def test_tsv_apply_weights(tmpdir, regtest):

    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True)

    _run_cmdline("pyprophet score --pi0_method=smoother --pi0_lambda 0.4 0 0 --in=test_data.txt --apply_weights=test_data_weights.csv "
                          "--test --ss_iteration_fdr=0.02")

def test_osw_0(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, pi0_lambda="0 0 0")

def test_osw_1(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, True, pi0_lambda="0 0 0")

def test_osw_2(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0")

def test_osw_3(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True)

def test_osw_4(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True, xgboost=True)

def test_osw_5(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True, xgboost=True, xgboost_tune=True)

def test_osw_6(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", score_filter=True)

def test_osw_7(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True, score_filter=True)

def test_osw_8(tmpdir, regtest):
  _run_pyprophet_osw_to_learn_model_metabo(regtest, tmpdir.strpath, True, ms1ms2=False, score_filter=False)

def test_osw_9(tmpdir, regtest):
  _run_pyprophet_osw_to_learn_model_metabo(regtest, tmpdir.strpath, True, ms1ms2=False, score_filter=True)

def test_osw_10(tmpdir,regtest):
  _run_pyprophet_osw_to_learn_model_metabo(regtest, tmpdir.strpath, True, ms1ms2=True, score_filter=True)

def test_not_unique_tg_id_blocks(tmpdir):

    os.chdir(tmpdir.strpath)
    data_path = os.path.join(DATA_FOLDER, "test_invalid_data.txt")
    shutil.copy(data_path, tmpdir.strpath)
    cmdline = "pyprophet score --pi0_method=smoother --pi0_lambda 0.4 0 0 --in=test_invalid_data.txt --test --ss_iteration_fdr=0.02"

    with pytest.raises(subprocess.CalledProcessError) as exc_info:
        subprocess.check_output(cmdline, shell=True, stderr=subprocess.STDOUT)

    e = exc_info.value
    assert "Error: group_id values do not form unique blocks in input file(s)." in str(e.output)
