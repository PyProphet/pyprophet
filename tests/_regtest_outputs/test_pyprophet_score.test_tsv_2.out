   cutoff     fdr       fn     fnr       fp     fpr     pep  pvalue  qvalue  svalue       tn        tp
0  3.0866  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026    0.00  0.9212  25.5215  332.9339
1  1.0884  0.0086  -3.4139  0.0000   3.1737  0.1240  0.3589  0.1240    0.01  1.0000  22.4139  364.8263
2  0.6049  0.0156  -5.7692  0.0000   5.8184  0.2274  1.0000  0.2274    0.02  1.0000  19.7692  367.1816
3 -0.6061  0.0507  -1.2151  0.0000  19.3725  0.7571  1.0000  0.7571    0.05  1.0000   6.2151  362.6275
4 -2.4409  0.0656  -0.1984  0.0000  25.3893  0.9922  1.0000  0.9922    0.10  1.0000   0.1984  361.6107
5     NaN     NaN      NaN     NaN      NaN     NaN     NaN     NaN    0.20     NaN      NaN       NaN
6     NaN     NaN      NaN     NaN      NaN     NaN     NaN     NaN    0.30     NaN      NaN       NaN
7     NaN     NaN      NaN     NaN      NaN     NaN     NaN     NaN    0.40     NaN      NaN       NaN
8     NaN     NaN      NaN     NaN      NaN     NaN     NaN     NaN    0.50     NaN      NaN       NaN
    cutoff     fdr       fn     fnr       fp     fpr     pep  pvalue  qvalue  svalue       tn        tp
0  -2.9318  0.0656  -0.1984  0.0000  25.3893  0.9922  1.0000  0.9922  0.0656  1.0000   0.1984  361.6107
1  -2.7158  0.0656  -0.1984  0.0000  25.3893  0.9922  1.0000  0.9922  0.0656  1.0000   0.1984  361.6107
2  -2.4998  0.0656  -0.1984  0.0000  25.3893  0.9922  1.0000  0.9922  0.0656  1.0000   0.1984  361.6107
3  -2.2838  0.0656  -0.1984  0.0000  25.3893  0.9922  1.0000  0.9922  0.0656  1.0000   0.1984  361.6107
4  -2.0678  0.0656  -0.1984  0.0000  25.3893  0.9922  1.0000  0.9922  0.0656  1.0000   0.1984  361.6107
5  -1.8518  0.0620  -0.6529  0.0000  23.9347  0.9354  1.0000  0.9354  0.0620  1.0000   1.6529  362.0653
6  -1.6358  0.0620  -0.6529  0.0000  23.9347  0.9354  1.0000  0.9354  0.0620  1.0000   1.6529  362.0653
7  -1.4198  0.0620  -0.6529  0.0000  23.9347  0.9354  1.0000  0.9354  0.0620  1.0000   1.6529  362.0653
8  -1.2038  0.0611  -0.0497  0.0000  23.5380  0.9199  1.0000  0.9199  0.0611  1.0000   2.0497  361.4620
9  -0.9878  0.0587  -0.0414  0.0000  22.5462  0.8811  1.0000  0.8811  0.0587  1.0000   3.0414  361.4538
10 -0.7718  0.0575   0.4296  0.1074  22.0172  0.8605  1.0000  0.8605  0.0575  1.0000   3.5704  360.9828
11 -0.5558  0.0507  -1.2151  0.0000  19.3725  0.7571  1.0000  0.7571  0.0507  1.0000   6.2151  362.6275
12 -0.3398  0.0448   0.3386  0.0376  16.9262  0.6615  1.0000  0.6615  0.0448  1.0000   8.6614  361.0738
13 -0.1238  0.0352  -1.3640  0.0000  13.2236  0.5168  1.0000  0.5168  0.0352  1.0000  12.3640  362.7764
14  0.0923  0.0331  -0.2236  0.0000  12.3640  0.4832  1.0000  0.4832  0.0331  1.0000  13.2236  361.6360
15  0.3083  0.0156  -5.7692  0.0000   5.8184  0.2274  1.0000  0.2274  0.0156  1.0000  19.7692  367.1816
16  0.5243  0.0156  -5.7692  0.0000   5.8184  0.2274  1.0000  0.2274  0.0156  1.0000  19.7692  367.1816
17  0.7403  0.0135  -5.5626  0.0000   5.0250  0.1964  0.7397  0.1964  0.0135  1.0000  20.5626  366.9750
18  0.9563  0.0114  -4.3561  0.0000   4.2315  0.1654  0.5454  0.1654  0.0114  1.0000  21.3561  365.7685
19  1.1723  0.0079  -2.6784  0.0000   2.9092  0.1137  0.3210  0.1137  0.0079  1.0000  22.6784  364.0908
20  1.3883  0.0054  -1.6041  0.0000   1.9835  0.0775  0.2088  0.0775  0.0054  1.0000  23.6041  363.0165
21  1.6043  0.0053   1.3298  0.0532   1.9174  0.0749  0.2018  0.0749  0.0053  0.9963  23.6702  360.0826
22  1.8203  0.0033   5.6025  0.1868   1.1901  0.0465  0.1282  0.0465  0.0033  0.9845  24.3975  355.8099
23  2.0363  0.0019   7.0736  0.2210   0.6612  0.0258  0.0549  0.0258  0.0019  0.9804  24.9264  354.3388
24  2.2523  0.0017   9.0075  0.2649   0.5951  0.0233  0.0435  0.0233  0.0017  0.9751  24.9925  352.4049
25  2.4683  0.0013  10.8752  0.3021   0.4628  0.0181  0.0228  0.0181  0.0013  0.9699  25.1248  350.5372
26  2.6843  0.0008  15.6769  0.3824   0.2645  0.0103  0.0049  0.0103  0.0008  0.9566  25.3231  345.7355
27  2.9003  0.0004  16.5446  0.3939   0.1322  0.0052  0.0011  0.0052  0.0004  0.9542  25.4554  344.8678
28  3.1163  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
29  3.3323  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
30  3.5483  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
31  3.7643  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
32  3.9803  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
33  4.1964  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
34  4.4124  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
35  4.6284  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
36  4.8444  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
37  5.0604  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
38  5.2764  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
39  5.4924  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
40  5.7084  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
41  5.9244  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
42  6.1404  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
43  6.3564  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
44  6.5724  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
45  6.7884  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
46  7.0044  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
47  7.2204  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
48  7.4364  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
49  7.6524  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
50  7.8684  0.0003  28.4785  0.5274   0.0661  0.0026  0.0005  0.0026  0.0003  0.9212  25.5215  332.9339
    d_score  decoy  group_id  main_var_xx_swath_prelim_score  p_value  peak_group_rank     pep  q_value  r_score  run_id  var_bseries_score  var_elution_model_fit_score  var_intensity_score  var_isotope_correlation_score  var_isotope_overlap_score  var_library_corr  var_library_rmsd  var_log_sn_score  var_massdev_score  var_massdev_score_weighted  var_norm_rt_score  var_xcorr_coelution  var_xcorr_coelution_weighted  var_xcorr_shape  var_xcorr_shape_weighted  var_yseries_score
0    5.1547      0  459_run0                          5.2789   0.0026                1  0.0005   0.0003   6.0024       0                  5                       0.9931               0.9305                         0.9987                     0.0000            0.9124            0.0399            5.0166             8.8392                      3.8845             0.0023               0.0000                        0.0000           0.9604                    0.9710                  6
1   -1.3725      0  459_run0                         -0.1397   0.9354                2  1.0000   0.0620   0.5231       0                  0                       0.9539               0.0259                         0.6325                     0.0000           -0.2855            0.1939            2.7888            27.0803                     21.2037             0.0107               1.3967                        0.8592           0.7892                    0.7164                  0
2   -4.3197      0  459_run0                         -0.7750   0.9922                7  1.0000   0.0656  -1.9510       0                  0                       0.6667               0.0004                         0.6579                     0.0000            0.3765            0.1117            0.5299             8.7979                      8.3272             0.0601               2.3868                        1.9700           0.1190                    0.0150                  0
3   -2.4375      0  459_run0                         -1.0545   0.9922                4  1.0000   0.0656  -0.3709       0                  0                       0.8473               0.0009                         0.7984                     0.2553            0.5187            0.0881            0.5683            15.3263                     17.4562             0.0200               4.3932                        2.7893           0.3292                    0.3308                  0
4   -1.6028      0  459_run0                         -1.2462   0.9354                3  1.0000   0.0620   0.3297       0                  0                      -1.0000               0.0004                         0.5638                     0.0455            0.4459            0.1497            0.5420            13.6321                     17.1255             0.0515               1.0000                        1.0000           0.0000                    0.0000                  0
..      ...    ...       ...                             ...      ...              ...     ...      ...      ...     ...                ...                          ...                  ...                            ...                        ...               ...               ...               ...                ...                         ...                ...                  ...                           ...              ...                       ...                ...
95  -3.0650      0  168_run0                         -1.0780   0.9922                4  1.0000   0.0656  -0.8978       0                  1                      -0.2038               0.0028                         0.9669                     0.0000           -0.6926            0.3447            0.0000             0.0604                      0.0381             0.0737               3.5921                        2.5829           0.1919                    0.1538                  0
96  -3.5587      0  168_run0                         -1.0963   0.9922                7  1.0000   0.0656  -1.3122       0                  0                       1.0000               0.0012                         0.8673                     0.0000           -0.6748            0.2056            0.0000            10.1516                     12.0983             0.0327               2.2331                        1.6284           0.3333                    0.1858                  0
97  -2.7735      0  168_run0                         -1.4446   0.9922                2  1.0000   0.0656  -0.6530       0                  0                       0.9411               0.0012                         0.0797                     0.0000           -0.4017            0.1385            0.0000            14.8865                     13.4708             0.0586               2.0539                        0.9411           0.7334                    0.6821                  0
98  -3.0728      0  168_run0                         -1.4526   0.9922                5  1.0000   0.0656  -0.9043       0                  1                       0.1498               0.0164                         0.9644                     0.0000           -0.7028            0.3422            0.2301             1.3878                      0.8742             0.0664               7.7597                        5.5046           0.3111                    0.1785                  0
99  -2.8249      0  168_run0                         -1.4706   0.9922                3  1.0000   0.0656  -0.6962       0                  1                       0.1509               0.0053                         0.6996                     0.0000            0.0856            0.2511            0.0000            10.4594                      9.8099             0.0235               4.5054                        2.9492           0.3262                    0.2549                  0

[100 rows x 26 columns]
   level                           score   weight
0    ms2  main_var_xx_swath_prelim_score   0.1554
1    ms2               var_bseries_score  -0.1381
2    ms2     var_elution_model_fit_score  -2.3916
3    ms2             var_intensity_score   0.8477
4    ms2   var_isotope_correlation_score   2.8853
5    ms2       var_isotope_overlap_score  -1.3842
6    ms2                var_library_corr  -0.7260
7    ms2                var_library_rmsd -13.8757
8    ms2                var_log_sn_score  -0.1271
9    ms2               var_massdev_score   0.0054
10   ms2      var_massdev_score_weighted  -0.0283
11   ms2               var_norm_rt_score -11.5622
12   ms2             var_xcorr_coelution   0.1361
13   ms2    var_xcorr_coelution_weighted  -0.0913
14   ms2                 var_xcorr_shape   3.9496
15   ms2        var_xcorr_shape_weighted   1.6273
16   ms2               var_yseries_score   0.1909
